go/updater/signature.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package updater
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/agl/ed25519"
+	"github.com/keybase/client/go/protocol"
+	"github.com/keybase/client/go/saltpack"
+	"github.com/keybase/client/go/saltpack/basic"
+)
+
+func SignArmoredDetachedFile(inFile string, signer saltpack.SigningSecretKey) (string, error) {
+	in, err := os.Open(inFile)
+	if err != nil {
+		return "", err
+	}
+	defer in.Close()
+	return SignArmoredDetached(in, signer)
+}
+
+func SignArmoredDetached(in io.Reader, signer saltpack.SigningSecretKey) (string, error) {
+	var buf bytes.Buffer
+	stream, err := saltpack.NewSignDetachedArmor62Stream(&buf, signer, "KEYBASE")
+	if err != nil {
+		return "", err
+	}
+	if _, err := io.Copy(stream, in); err != nil {
+		return "", err
+	}
+	if err := stream.Close(); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func VerifyUpdate(update keybase1.Update, publicKeyHex string) error {
+	publicKey, err := decodePublicKeyHex(publicKeyHex)
+	if err != nil {
+		return err
+	}
+
+	signature := update.Asset.Signature
+
+	err = VerifyArmoredDetachedFile(update.Asset.LocalPath, signature, publicKey)
+	if err != nil {
+		return fmt.Errorf("Couldn't verify update download asset: %s", err)
+	}
+	return nil
+}
+
+func VerifyArmoredDetachedFile(inFile string, signature string, publicKey *[32]byte) error {
+	in, err := os.Open(inFile)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	return VerifyArmoredDetached(in, signature, publicKey)
+}
+
+func VerifyArmoredDetached(reader io.Reader, signature string, publicKey *[32]byte) error {
+	signingPublicKey := basic.NewSigningPublicKey(publicKey)
+	publicKeyFromSig, brand, err := saltpack.Dearmor62VerifyDetachedReader(reader, signature, signingPublicKey)
+	if err != nil {
+		return err
+	}
+	if brand != "KEYBASE" {
+		return fmt.Errorf("Invalid brand")
+	}
+	if !saltpack.PublicKeyEqual(publicKeyFromSig, signingPublicKey) {
```

Comment:
We should be able to accommodate several signing keys here. This will be important for our ability to cycle out old keys and to introduce new ones.